firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+
+    WHEN("transform is called on three parseable datagram bodies, whose headers have sequence numbers as 1, 10 & 11 respectively")
+      THEN("The first transform method reports ok status")
+      THEN("After the first transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+      THEN("The second transform method reports ok status")
+      THEN("After the second transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+      THEN("The third transform method reports ok status")
+      THEN("After the third transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+  GIVEN("A Datagram receiver of buffer size 254 bytes and expected sequence number equal to 0xff")
+  // assertions carried out in GIVEN:
+  // The output_datagram's seq accessor method returns 0xff
+  // The value returned by the output_datagram's length accessor method is 8
+  // The payload buffer returned by the output_datagram's payload accessor method is '0xff 0x06 0xcd 0x6a 0xc2 0x7f 0xa1 0x5b'  
+
+    WHEN("transform is called on two parseable datagram bodies, whose headers have sequence numbers as 0 & 1 respectively")
```

Comment:
The DatagramReceiver is expecting its next datagram to have sequence number 0xff. In this WHEN, do you mean that the headers will have sequence numbers 0xff and 0x00?

I'd also suggest adding a WHEN for a header with sequence 0xff and any non-0x00 value, so we can guarantee that 0xff isn't a special case where any sequence number is accepted for the next datagram.

Reply:
> The DatagramReceiver is expecting its next datagram to have sequence number 0xff. In this WHEN, do you mean that the headers will have sequence numbers 0xff and 0x00?
yep thanks for pointing it out, I did a derp in setting up the state before the WHEN. yes I meant 0xff  and 0x00

> I'd also suggest adding a WHEN for a header with sequence 0xff and any non-0x00 value, so we can guarantee that 0xff isn't a special case where any sequence number is accepted for the next datagram.

will do.